system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -69,6 +70,93 @@ class TopSite extends React.Component {
   }
 }
 
+/**
+ * A proxy class that uses double requestAnimationFrame from
+ * componentDidMount to dispatch a SAVE_SESSION_PERF_DATA to the main procsess
+ * after the paint.
+ *
+ * This uses two callbacks because, after one callback, this part of the tree
+ * may have rendered but not yet reflowed.  This strategy is modeled after
+ * https://stackoverflow.com/a/34999925 but uses a double rFA because
+ * we want to get to the closest reliable paint for measuring, and
+ * setTimeout is often throttled or queued by browsers in ways that could
+ * make it lag too long.
+ *
+ * XXX Should probably be split out into a higher-order component to wrap
+ * whatever.
+ *
+ * @class TopSitesPerfTimer
+ * @extends {React.Component}
+ */
+class TopSitesPerfTimer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.perfSvc = this.props.perfSvc || perfSvc;
+
+    this._maybeSendPaintedEvent = this._maybeSendPaintedEvent.bind(this);
+    this._sendPaintedEvent = this._sendPaintedEvent.bind(this);
+    this._timestampSent = false;
+  }
+
+  componentDidMount() {
+    this._maybeSendPaintedEvent();
+  }
+
+  componentDidUpdate() {
+    this._maybeSendPaintedEvent();
+  }
+
+  /**
+   * Call the given callback when the next animation frame (not the upcoming
+   * one) paints.
+   *
+   * @param {Function} callback
+   *
+   * @returns void
+   */
+  _onNextFrame(callback) {
+    requestAnimationFrame(() => {
+      requestAnimationFrame(callback);
+    });
+  }
+
+  _maybeSendPaintedEvent() {
+    // If we've already saved a timestamp for this session, don't do so again.
+    if (this._timestampSent) {
+      return;
+    }
+
+    // We don't want this to ever happen, but sometimes it does.  And when it
+    // does (typically on the first newtab at startup time calling
+    // componentDidMount), the paint(s) we care about will be later (eg
+    // in a subsequent componentDidUpdate).
+    if (!this.props.TopSites.initialized) {
```

Comment:
No, because that doesn't mean that the remained of the TopSites have mounted, I don't think.  Even if it does, given our current timelines, I don't think it's worth the time to move things around, because it doesn't really make a big difference in code readability/maintainability.